src/SerialIO.c

```diff
@@ -0,0 +1,58 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "SerialIO.h"
+#include <stdio.h>
+#include "Serial.h"
+
+static int
+Print(uint32_t aLength, void* aBuffer)
+{
+  IPCMessageQueue* queue = GetSerialOutQueue();
+  if (!queue) {
+    return -1;
+  }
+
+  IPCMessage msg;
+  int res = IPCMessageInit(&msg);
+  if (res < 0) {
+    return -1;
+  }
+  res = IPCMessageProduce(&msg, aLength, aBuffer);
+  if (res < 0) {
+    return -1;
+  }
+  res = IPCMessageQueueConsume(queue, &msg);
+  if (res < 0) {
+    return -1;
+  }
+  return IPCMessageWaitForConsumption(&msg);
+}
+
+/*
+ * Libc-like interfaces for easy usage.
+ */
+
+int
+print(const char* fmt, ...)
+{
+  va_list ap;
+
+  va_start(ap, fmt);
+  int res = vprint(fmt, ap);
+  va_end(ap);
+
+  return res;
+}
+
+int
+vprint(const char* fmt, va_list ap)
+{
+  char buf[128];
+  int res = vsnprintf(buf, sizeof(buf), fmt, ap);
+  if (res < 0) {
+    return -1;
+  }
+  return Print(sizeof(buf), buf);
```

Comment:
For printing, I think its a waste to have a printing task, especially since this is synchronous. I really like the way that the linux kernel printk works in that it can be called from within an ISR or at thread level. This print function can only be called from thread level.